Adding a get_slot_fields_and_types method to python msg classes

mlautman opened the pull request:
This adds a `_slot_types` field to python message classes as was requested here: https://github.com/ros2/rosidl_python/issues/16

This attempts to remain as true to the ROS1 functionality as possible

mlautman commented:
@dirk-thomas

dirk-thomas commented on rosidl_generator_py/resource/_msg.py.em:
```diff
@@ -114,9 +114,27 @@ class @(spec.base_type.type)(metaclass=Metaclass):
+@[for field in spec.fields]@
+@[  if field.type.is_primitive_type() ]@
+@[    if field.type.is_array ]@
+        '@(get_python_type(field.type))[@(field.type.array_size)]',
```
This doesn't distinguish between a static array (with a fixed size) and a dynamic array (either with an upper bound or without a bound).

mlautman replied:
True. Do you have a preferred method to denote the different types of arrays? I didn't have a strong intuition here

dirk-thomas replied:
I would suggest either the same notation as in the `.msg` or an object representation. Thinking about it the second one is probably be better in terms of stability with the upcoming change to `.idl` files since the notation will change (see https://github.com/ros2/design/blob/gh-pages/articles/142_idl.md).

mlautman replied:
I know I can use `field.type.array_size.is_upper_bound` for the upper bound arrays but I'm not sure how to check for unbounded arrays. Any hints?

dirk-thomas replied:
Looking at the patch I don't see where the type is being distinguished?

mlautman replied:
My apologies, I got my branches mixed up. All of my changes are in one commit (b9a6a81abc8f75fec9899eb0e9c71b889b448620) but I'm unsure of whether I should cherry pick that commit on top of master or 0.5.2. Since this is targeting master, I assume that I should cherry pick on top of that, but doing so causes the build to fail for unrelated reasons..  

`/home/mike/ws_rqt2/build/rosidl_generator_py/rosidl_generator_py/rosidl_generator_py/msg/_string_arrays_s.c:12:10: fatal error: rosidl_generator_c/primitives_sequence.h: No such file or directory
 #include <rosidl_generator_c/primitives_sequence.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
/home/mike/ws_rqt2/build/rosidl_generator_py/rosidl_generator_py/rosidl_generator_py/msg/_nested_s.c:12:10: fatal error: rosidl_generator_c/primitives_sequence.h: No such file or directory
 #include <rosidl_generator_c/primitives_sequence.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
/home/mike/ws_rqt2/build/rosidl_generator_py/rosidl_generator_py/rosidl_generator_py/msg/_various_s.c:12:10: fatal error: rosidl_generator_c/primitives_sequence.h: No such file or directory
 #include <rosidl_generator_c/primitives_sequence.h>
`

Is there a .repos available with the most recent ros2 branches that I should use for making PR's to master?

mlautman replied:
@dirk-thomas

dirk-thomas replied:
It depends on what you are trying to build. I would suggest using the `.repos` files from the `master` branch in the `ros2/ros2` repo and building everything from source for your use case. Then a patch based on the master of this repo will work.

(The opposite would be to use the `bouncy` branch of the `.repos` file and then also target the `bouncy` branch in this repo. But usually all contributions are targeting the default branch and only in case of severe bug fixes are being considered for backporting afterwards.)

dirk-thomas replied:
The current state seems to pass on CI: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5548)](https://ci.ros2.org/job/ci_linux/5548/)

mlautman commented:
@dirk-thomas ready for another review

dirk-thomas commented on rosidl_generator_py/resource/_msg.py.em:
```diff
@@ -114,9 +114,31 @@ class @(spec.base_type.type)(metaclass=Metaclass):
+    _slot_types = [
+@[for field in spec.fields]@
+@[  if field.type.is_primitive_type() ]@
+        '@(field.type)',
```
A string could still have an upper boundary.

dirk-thomas replied:
The string representation of primitives contains these parts already.

mlautman commented:
@dirk-thomas Changed `_slot_types_dict` to `get_slot_types_dict()` and added noqa W504 where appropriate to pass flake8

mlautman commented:
https://ci.ros2.org/job/ci_linux/5564/

dirk-thomas commented:
Ignoring W504 all over the place isn't necessary. This has been addressed in the linter configuration already a while ago: ament/ament_lint#110.

mlautman commented:
@dirk-thomas It still fails locally with `colcon build test --package-select rosidl_python` without that change

dirk-thomas commented:
> It still fails locally

Are you using the latest default branches of all the other ROS 2 repos?

mlautman commented on rosidl_generator_py/resource/_msg.py.em:
```diff
@@ -175,6 +197,10 @@ class @(spec.base_type.type)(metaclass=Metaclass):
         return True
+
+    @@classmethod
+    def get_slot_types_dict(cls):
```
```suggestion
    def get_slot_fields_and_types(cls):
```

brawner replied:
```suggestion
    def get_fields_and_field_types(cls):
```

mlautman commented on rosidl_generator_py/test/test_interfaces.py:
```diff
@@ -257,3 +257,51 @@ def test_out_of_range():
+    a = Nested()
+    assert hasattr(a, 'get_slot_types_dict')
+    assert hasattr(a, '__slots__')
+    nested_slot_types_dict = getattr(a, 'get_slot_types_dict')()
```
```suggestion
    nested_slot_types_dict = getattr(a, 'get_slot_fields_and_types')()
```

mlautman commented on rosidl_generator_py/test/test_interfaces.py:
```diff
@@ -257,3 +257,51 @@ def test_out_of_range():
+
+def test_slot_attributes():
+    a = Nested()
+    assert hasattr(a, 'get_slot_types_dict')
```
```suggestion
    assert hasattr(a, 'get_slot_fields_and_types')
```

mlautman commented on rosidl_generator_py/test/test_interfaces.py:
```diff
@@ -257,3 +257,51 @@ def test_out_of_range():
+
+def test_primative_slot_attributes():
+    b = StringArrays()
+    assert hasattr(b, 'get_slot_types_dict')
```
```suggestion
    assert hasattr(b, 'get_slot_fields_and_types')
```

mlautman commented on rosidl_generator_py/test/test_interfaces.py:
```diff
@@ -257,3 +257,51 @@ def test_out_of_range():
+    b = StringArrays()
+    assert hasattr(b, 'get_slot_types_dict')
+    assert hasattr(b, '__slots__')
+    string_slot_types_dict = getattr(b, 'get_slot_types_dict')()
```
```suggestion
    string_slot_types_dict = getattr(b, 'get_slot_fields_and_types')()
```

brawner commented on rosidl_generator_py/resource/_msg.py.em:
```diff
@@ -175,6 +197,10 @@ class @(spec.base_type.type)(metaclass=Metaclass):
+
+    @@classmethod
+    def get_fields_and_field_types(cls):
+        return cls._slot_types_dict
```
Since this is a getter, should it return a shallow copy so that you can modify the underlying dictionary?

mlautman replied:
Good catch. Fixed.

mlautman commented:
@dirk-thomas I am using the ros2.repos that you passed along to us with the 0.5.2 release.

I went back and removed W504. I also changed the code to return a copy of the dict rather than the dict itself per @brawner's comment

mlautman commented on rosidl_generator_py/resource/_msg.py.em:
```diff
@@ -252,7 +279,7 @@ bound = 2**nbits
                 "The '@(field.name)' field must be @(' '.join(assert_msg_suffixes))"
 @[  elif field.type.string_upper_bound]@
-                ((isinstance(value, str) or isinstance(value, UserString)) and
+                ((isinstance(value, str) or isinstance(value, UserString)) and  # noqa W504
```
```suggestion
                ((isinstance(value, str) or isinstance(value, UserString)) and
```

dirk-thomas commented:
This can be merged as is for now. With the coming change I will likely change this in the following way:

* Store the types in a class constant, e.g. `SLOT_TYPES`.
* Change the data structure from a mutable `dict` to an immutable `tuple` of [abstract types](https://github.com/ros2/rosidl/blob/49b1eaa23eb4d11c73dca7ea610a1cce6f129607/rosidl_parser/rosidl_parser/definition.py#L48)
* Either remove the class method (or rename it to `get_slot_types` which zips `__slots__` and `SLOT_TYPES` and returns an ordered dictionary).
